utils/checks.stanza

```diff
@@ -113,8 +133,207 @@ public pcb-check rated-temperature (component:JITXObject):
     locators =             [component]  
   )
 
-public pcb-check io (module:JITXObject) :
+; Returns all the pins that are electrically-meaningful, ie are not 
+; from components marked do-not-place
+defn all-pins (module:JITXObject) -> Tuple<JITXObject> : 
+  populated-components(module) 
+    $> seq-cat{pins, _}
+    $> to-tuple
+  
+; Checks if a pin is marked "no-connect"
+defn no-connect? (p:JITXObject) -> True|False : 
+  has-property?(p.no-connect) or
+  has-property?(p.nc) or 
+  has-property?(p.NC)
+
+; Returns if a pin is a digital-io pin
+defn is-digital-io-pin? (p:JITXObject) -> True|False : 
+  has-property?(p.digital-io) or 
+  has-property?(p.digital-output)
+
+; Returns if a pin is a digital-input pin
+defn is-digital-input? (p:JITXObject) -> True|False : 
+  has-property?(p.digital-io) or 
+  has-property?(p.digital-input)
+
+; A PCB check that walks a module and verifies if all pins
+; marked no-connect are indeed not connected
+;
+pcb-check no-connect-pins-not-connected (module:JITXObject) : 
+  val all-pins = all-pins(module)
+  val nc-pins = filter(no-connect?, all-pins)
+  for p in nc-pins do : 
+    val connected? = any?(connected?{[_, p]}, all-pins)
+    #CHECK(
+      condition =            not connected?
+      name =                 "IO pin checks"
+      description =          "Check io pin properties"
+      category =             "pin-check"
+      subcheck-description = "Check that no-connect pads are not connected",
+      pass-message =         "Pin %_ is marked no-connect and is correctly not connected" % [ref(p)],
+      fail-message =         "Pin %_ is marked no-connect and is incorrectly connected to other pins" % [ref(p)],
+      locators =             [p])
+
+; A check used to verify the electrical properties of 
+; digital i/o pins.
+pcb-check digital-io-levels (module:JITXObject) : 
+  val all-pins = all-pins(module)
+  val io-pins  = filter(is-digital-io-pin?, all-pins)
+  for p in io-pins do : 
+    val all-connected-pins = all-pins 
+      $> filter{{connected?([p, _0]) and _0 != p}, _}
+      $> to-tuple   
+    val connected-io-pins = all-connected-pins
+      $> filter{is-digital-io-pin?, _}
+      $> to-tuple
+    
+    #CHECK(condition =            not empty?(connected-io-pins),
```

Comment:
This check would seem to check if digital-output and digital-io pins are positively connected to at least one other digital-io pin. That check is not quite what we want as we should check and error only when more than 1 digital-output (not digital-io) pin is connected together. You can have a digital-output and digital-io pin 
1) unconnected
2) connected to each other (digital-output -> 1+ digital-io pins)
3) connected to 1+ digital-input pis

Reply:
Can you clarify the 2nd case, is that multiple digital output pins connected to each other? I'm not sure if this check would ever fail if the pin can be unconnected or connected to any number of digital output/input/io pins. Would we fail if it was connected to any pins that are *not* digital i/o instead? 

This check was translated directly from the current implementation of `io`, the logic doesn't seem correct to me in either case (input pins are never checked at this stage).

Reply:
More than one digital-output pin connected together is a no-no.
One digital-output can connect to 1 or more digital-io pins (because they can be used as input). 
One digital-output can be unconnected (and that is not an error)
Digital-output cannot be connected to power-supply-pins.

We do have to enhance the checks a bit which I want to do with your assistance (like checking digital-input pins)

Reply:
Ok, updated. The check is slightly more specific now : 

- All i/o pins are checked (using the function you described below)
- If the i/o pin is an output, check that it isn't connected to any other outputs

Let me know if this meets what we're after, I think the semantics of the original check are being changed slightly